jsonrpc/debug_endpoint.go

```diff
@@ -0,0 +1,359 @@
+	filter BlockNumberOrHash,
+	config *TraceConfig,
+) (interface{}, error) {
+	var (
+		header *types.Header
+		err    error
+	)
+
+	// The filter is empty, use the latest block by default
+	if filter.BlockNumber == nil && filter.BlockHash == nil {
+		filter.BlockNumber, _ = createBlockNumberPointer("latest")
+	}
+
+	header, err = d.getHeaderFromBlockNumberOrHash(&filter)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get header from block hash or block number")
+	}
+
+	tx, err := d.decodeTxn(arg)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the caller didn't supply the gas limit in the message, then we set it to maximum possible => block gas limit
+	if tx.Gas == 0 {
+		tx.Gas = header.GasLimit
+	}
+
+	tracer := tracer.NewStructTracer()
+
+	return d.store.TraceCall(tx, header, tracer)
+}
+
+func (d *Debug) traceBlock(
+	block *types.Block,
+	config *TraceConfig,
+) (interface{}, error) {
+	if block.Number() == 0 {
+		return nil, errors.New("genesis is not traceable")
+	}
+
+	tracer := tracer.NewStructTracer()
+
+	return d.store.TraceMinedBlock(block, tracer)
+}
+
+func (d *Debug) getTxAndBlockByTxHash(txHash types.Hash) (*types.Transaction, *types.Block) {
+	blockHash, ok := d.store.ReadTxLookup(txHash)
+	if !ok {
+		return nil, nil
+	}
+
+	block, ok := d.store.GetBlockByHash(blockHash, true)
+	if !ok {
+		return nil, nil
+	}
+
+	for _, txn := range block.Transactions {
+		if txn.Hash == txHash {
+			return txn, block
+		}
+	}
+
+	return nil, nil
+}
+
+// getNumericBlockNumber returns block number based on current state or specified number
+func (d *Debug) getNumericBlockNumber(number BlockNumber) (uint64, error) {
+	switch number {
+	case LatestBlockNumber:
+		return d.store.Header().Number, nil
+
+	case EarliestBlockNumber:
+		return 0, nil
+
+	case PendingBlockNumber:
+		return 0, fmt.Errorf("fetching the pending header is not supported")
+
+	default:
+		if number < 0 {
+			return 0, fmt.Errorf("invalid argument 0: block number larger than int64")
+		}
+
+		return uint64(number), nil
+	}
+}
+
+// getBlockHeader returns a header using the provided number
+func (d *Debug) getBlockHeader(number BlockNumber) (*types.Header, error) {
+	switch number {
+	case LatestBlockNumber:
+		return d.store.Header(), nil
+
+	case EarliestBlockNumber:
+		header, ok := d.store.GetHeaderByNumber(uint64(0))
+		if !ok {
+			return nil, fmt.Errorf("error fetching genesis block header")
+		}
+
+		return header, nil
+
+	case PendingBlockNumber:
+		return nil, fmt.Errorf("fetching the pending header is not supported")
+
+	default:
+		// Convert the block number from hex to uint64
+		header, ok := d.store.GetHeaderByNumber(uint64(number))
+		if !ok {
+			return nil, fmt.Errorf("error fetching block number %d header", uint64(number))
+		}
+
+		return header, nil
+	}
+}
+
+// getHeaderFromBlockNumberOrHash returns a header using the provided number or hash
+func (d *Debug) getHeaderFromBlockNumberOrHash(bnh *BlockNumberOrHash) (*types.Header, error) {
```

Comment:
Many of the helper methods have been copied from `eth` endpoint. What are your thoughts on integrating that?

Reply:
Firstly I had reverted creating common struct due to this discussion https://github.com/0xPolygon/polygon-edge/pull/836#discussion_r1009168456

Reply:
I added helper functions instead of super struct b9b3d8e6e242ad4041cc53e5460ed2fffc9aeb42